pkg/agent/controller/networkpolicy/node_reconciler_linux.go

```diff
@@ -0,0 +1,703 @@
+			}
+		}
+	} else {
+		// If these are new rules, append the new rules then sort all rules.
+		cachedIPTRules = append(cachedIPTRules, iptRules...)
+		sort.Slice(cachedIPTRules, func(i, j int) bool {
+			return !cachedIPTRules[i].priority.Less(*cachedIPTRules[j].priority)
+		})
+	}
+
+	// Get all the sorted iptables rules and synchronize them.
+	var iptRuleStrs []string
+	for _, r := range cachedIPTRules {
+		iptRuleStrs = append(iptRuleStrs, r.ruleStr)
+	}
+	if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPTables([]string{iptChain}, [][]string{iptRuleStrs}, isIPv6); err != nil {
+		return err
+	}
+
+	// cache the new iptables rules.
+	cachedCoreIPTChain.rules = cachedIPTRules
+	return nil
+}
+
+func (r *nodeReconciler) deleteCoreIPRule(ruleID string, iptChain string, isIPv6 bool) error {
+	cachedCoreIPTChain := r.getCachedCoreIPTChain(iptChain, isIPv6)
+	cachedCoreIPTChain.Lock()
+	defer cachedCoreIPTChain.Unlock()
+
+	// Get all cached iptables rules, then delete the rule with the given ruleID.
+	cachedIPTRules := cachedCoreIPTChain.rules
+	var indexToDelete int
+	for i := 0; i < len(cachedIPTRules); i++ {
+		if cachedIPTRules[i].ruleID == ruleID {
+			indexToDelete = i
+			break
+		}
+	}
+	cachedIPTRules = append(cachedIPTRules[:indexToDelete], cachedIPTRules[indexToDelete+1:]...)
+
+	// Get all the sorted iptables rules and synchronize them.
+	var iptRuleStrs []string
+	for _, r := range cachedIPTRules {
+		iptRuleStrs = append(iptRuleStrs, r.ruleStr)
+	}
+	if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPTables([]string{iptChain}, [][]string{iptRuleStrs}, isIPv6); err != nil {
+		return err
+	}
+
+	// cache the new iptables rules.
+	cachedCoreIPTChain.rules = cachedIPTRules
+	return nil
+}
+
+func (r *nodeReconciler) getCachedCoreIPTChain(iptChain string, isIPv6 bool) *coreIPTChain {
+	// There are 4 categories of cached core iptables rules:
+	// - For IPv4, iptables rules installed in chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv6, ip6tables rules installed in chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv4, iptables rules installed in chain ANTREA-EGRESS-RULES for egress rules.
+	// - For IPv6, ip6tables rules installed in chain ANTREA-EGRESS-RULES for egress rules.
+	categoryKey := genCacheCategory(iptChain, isIPv6)
+	return r.cachedCoreIPTChains[categoryKey]
+}
+
+func groupMembersToIPNets(groups v1beta2.GroupMemberSet, isIPv6 bool) sets.Set[string] {
+	ipNets := sets.New[string]()
+	suffix := "/32"
+	if isIPv6 {
+		suffix = "/128"
+	}
+	for _, member := range groups {
+		for _, ip := range member.IPs {
+			ipAddr := net.IP(ip)
+			if isIPv6 == utilnet.IsIPv6(ipAddr) {
+				ipNets.Insert(ipAddr.String() + suffix)
+			}
+		}
+	}
+	return ipNets
+}
+
+func ipBlocksToIPNets(ipBlocks []v1beta2.IPBlock, isIPv6 bool) sets.Set[string] {
+	ipNets := sets.New[string]()
+	for _, b := range ipBlocks {
+		blockCIDR := ip.IPNetToNetIPNet(&b.CIDR)
+		if isIPv6 != utilnet.IsIPv6CIDR(blockCIDR) {
```

Comment:
Ditto, if `isIPv6` is true, I suppose there will be no ipv4 CIDR in the ipBlocks?

Reply:
This is from `if isIPv6 && !utilnet.IsIPv6CIDR(blockCIDR) || !isIPv6 && utilnet.IsIPv6CIDR(blockCIDR)`